### PR TITLE
planner: throw error when insert values is correlated subquery

### PR DESCRIPTION
**Pull request by wshwsh12**
### What problem does this PR solve?

Issue Number: ref https://github.com/pingcap/tidb/issues/30626

Problem Summary:

### What is changed and how it works?

Now tidb doesn't support insert correlated subquery, so we should throw error to avoid undefine behavior(panic, crash or insert wrong result).

### Check List

Tests 

- [x] Unit test
- [ ] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No code

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

```release-note
None
```

### CODE REVIEW

File: planner/core/planbuilder.go [hunk 1]
```diff
@@ -3234,10 +3234,15 @@ func (b *PlanBuilder) buildSetValuesOfInsert(ctx context.Context, insert *ast.In
...
 			usingPlan = LogicalTableDual{}.Init(b.ctx, b.getSelectOffset())
 		}
-		expr, _, err := b.rewriteWithPreprocess(ctx, assign.Expr, usingPlan, nil, nil, true, checkRefColumn)
+		expr, np, err := b.rewriteWithPreprocess(ctx, assign.Expr, usingPlan, nil, nil, true, checkRefColumn)
 		if err != nil {
 			return err
 		}
+		if np != nil {
+			if _, ok := np.(*LogicalTableDual); !ok {
+				return errors.New("Insert's SET opration or VALUES_LIST doesn't support the subquery now")
```
Comment (winoros) [inline planner/core/planbuilder.go:3243]:
```suggestion
				return errors.New("Insert's SET opration or VALUES_LIST doesn't support complex subqueries now")
```

File: planner/core/planbuilder.go [hunk 2]
```diff
@@ -3314,7 +3319,13 @@ func (b *PlanBuilder) buildValuesListOfInsert(ctx context.Context, insert *ast.I
 				if _, ok := valueItem.(*ast.SubqueryExpr); ok {
 					usingPlan = LogicalTableDual{}.Init(b.ctx, b.getSelectOffset())
 				}
-				expr, _, err = b.rewriteWithPreprocess(ctx, valueItem, usingPlan, nil, nil, true, checkRefColumn)
+				var np LogicalPlan
+				expr, np, err = b.rewriteWithPreprocess(ctx, valueItem, usingPlan, nil, nil, true, checkRefColumn)
+				if np != nil {
+					if _, ok := np.(*LogicalTableDual); !ok {
+						return errors.New("Insert's SET opration or VALUES_LIST doesn't support the subquery now")
```
Comment (winoros) [inline planner/core/planbuilder.go:3326]: ditto

**Review by winoros — APPROVED** [review]

File: planner/core/planbuilder.go [hunk 3]
```diff
@@ -3234,10 +3234,15 @@ func (b *PlanBuilder) buildSetValuesOfInsert(ctx context.Context, insert *ast.In
...
 			usingPlan = LogicalTableDual{}.Init(b.ctx, b.getSelectOffset())
 		}
-		expr, _, err := b.rewriteWithPreprocess(ctx, assign.Expr, usingPlan, nil, nil, true, checkRefColumn)
+		expr, np, err := b.rewriteWithPreprocess(ctx, assign.Expr, usingPlan, nil, nil, true, checkRefColumn)
 		if err != nil {
 			return err
 		}
+		if np != nil {
+			if _, ok := np.(*LogicalTableDual); !ok {
+				return errors.New("Insert's SET operation or VALUES_LIST doesn't support complex subqueries now")
```
Comment (winoros) [inline planner/core/planbuilder.go:3243]:
```suggestion
				// See issue#30626 and the related tests in function TestInsertValuesWithSubQuery for more details.
				// This is a TODO and we will support it later.
				return errors.New("Insert's SET operation or VALUES_LIST doesn't support complex subqueries now")
```

**Review by AilinKid — APPROVED** [review]
LGTM

**Pull request merged**